src/PostgREST/App.hs

```diff
@@ -252,14 +253,21 @@ app dbStructure conf apiRequest =
                          singular paramsAsSingleObject
                          (contentType == CTTextCSV)
                          (contentType == CTOctetStream) _isReadOnly bField
-              let (tableTotal, queryTotal, body) =
-                    fromMaybe (Just 0, 0, "[]") row
+              let (tableTotal, queryTotal, body, jsonHeaders) =
+                    fromMaybe (Just 0, 0, "[]", Nothing) row
                   (status, contentRange) = rangeHeader queryTotal tableTotal
-              if singular && queryTotal /= 1
-                then do
-                  HT.condemn
-                  return $ singularityError (toInteger queryTotal)
-                else return $ responseLBS status [toHeader contentType, contentRange] (toS body)
+                  decodedHeaders = first toS <$> (eitherDecode <$> (toS <$> jsonHeaders)) :: Maybe (Either Text [GucHeader])
```

Comment:
Trying to understand this somewhat complex type. What's the difference between `Nothing`, `Left t` and `Right []`?

Reply:
If the user doesn't do a `set local "response.headers"` then this gives Nothing, this comes from the `Maybe ByteString` in `type ProcResults` in QueryBuilder.

`Left t` happens when the `response.headers` has bad format, like in `"response.headers" = '{"X-Test": "invalid"}'`(can only be array) or `"response.headers" = '{{}}'`. I tried to use aeson `fail` function in `parseJSON` to specify the error but sometimes it didn't respond with my custom error so that's why I don't use the `t` in `Left t`.  

`Just (Right [])` can only happen if `response.headers = '[]'` is set, I made this strict so even `response.headers='[{"Set-Cookie": null}]'/'[{"Set-Cookie": 666}]'/'[{"Set-Cookie": true}]'` will be a `Left`.

I struggled a bit to keep this part DRY, for now don't know how to make it clearer.

Reply:
What about treating two of these cases the same: when the user doesn't set `response.headers` and when they set it to `[]`? We could then simplify the type to `Either Text [GucHeader]` and return `Right []` when `response.headers` is not set.